File: src/main/java/storebackend/service/DhlLayoutService.java
package storebackend.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import storebackend.dto.DhlLayoutRequests.*;
import storebackend.dto.DhlShelfSlotLayoutDto;
import storebackend.dto.DhlZoneDto;
import storebackend.dto.DhlZoneRequest;
import storebackend.entity.*;
import storebackend.enums.DhlParcelStatus;
import storebackend.repository.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * DHL Layout Service (Phase 3A)
 * 
 * Business Logic für visuellen Regalplan:
 * - Zonen verwalten
 * - Layout speichern/laden
 * - Batch-Updates für Drag&Drop
 * - Validation
 * 
 * WICHTIG:
 * - DhlShelfSlot.id bleibt stabile Identität
 * - Layout-Änderungen verändern niemals Paketzuordnungen
 * - Multi-Tenant strikt validieren
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class DhlLayoutService {
    
    private final DhlZoneRepository zoneRepository;
    private final DhlShelfSlotLayoutRepository layoutRepository;
    private final DhlShelfSlotRepository slotRepository;
    private final DhlParcelRepository parcelRepository;
    private final StoreRepository storeRepository;

    // ========== ZONES ==========
    
    /**
     * Listet alle Zonen eines Stores
     */
    @Transactional(readOnly = true)
    public List<DhlZoneDto> getZones(Long storeId) {
        log.debug("Loading zones for store {}", storeId);
        return zoneRepository.findByStoreIdOrderBySortOrder(storeId).stream()
            .map(DhlZoneDto::fromEntity)
            .collect(Collectors.toList());
    }
    
    /**
     * Erstellt neue Zone
     */
    @Transactional
    public DhlZoneDto createZone(Long storeId, DhlZoneRequest request) {
        log.info("Creating zone '{}' for store {}", request.getName(), storeId);
        
        // Validation
        if (request.getName() == null || request.getName().trim().isEmpty()) {
            throw new IllegalArgumentException("Zone name is required");
        }
        
        // Check duplicate name
        if (zoneRepository.existsByStoreIdAndName(storeId, request.getName(), null)) {
            throw new IllegalArgumentException("Zone with name '" + request.getName() + "' already exists");
        }
        
        Store store = storeRepository.findById(storeId)
            .orElseThrow(() -> new IllegalArgumentException("Store not found"));
        
        DhlZone zone = new DhlZone();
        zone.setStore(store);
        zone.setName(request.getName());
        zone.setColor(request.getColor());
        zone.setSortOrder(request.getSortOrder() != null ? request.getSortOrder() : 0);
        
        zone = zoneRepository.save(zone);
        return DhlZoneDto.fromEntity(zone);
    }
    
    /**
     * Aktualisiert Zone
     */
    @Transactional
    public DhlZoneDto updateZone(Long storeId, Long zoneId, DhlZoneRequest request) {
        log.info("Updating zone {} for store {}", zoneId, storeId);
        
        DhlZone zone = zoneRepository.findByStoreIdAndId(storeId, zoneId)
            .orElseThrow(() -> new IllegalArgumentException("Zone not found or access denied"));
        
        // Validation
        if (request.getName() == null || request.getName().trim().isEmpty()) {
            throw new IllegalArgumentException("Zone name is required");
        }
        
        // Check duplicate name (exclude current zone)
        if (zoneRepository.existsByStoreIdAndName(storeId, request.getName(), zoneId)) {
            throw new IllegalArgumentException("Zone with name '" + request.getName() + "' already exists");
        }
        
        zone.setName(request.getName());
        zone.setColor(request.getColor());
        if (request.getSortOrder() != null) {
            zone.setSortOrder(request.getSortOrder());
        }
        
        zone = zoneRepository.save(zone);
        return DhlZoneDto.fromEntity(zone);
    }
    
    /**
     * Löscht Zone
     */
    @Transactional
    public void deleteZone(Long storeId, Long zoneId) {
        log.info("Deleting zone {} for store {}", zoneId, storeId);
        
        DhlZone zone = zoneRepository.findByStoreIdAndId(storeId, zoneId)
            .orElseThrow(() -> new IllegalArgumentException("Zone not found or access denied"));
        
        // Note: Layouts referencing this zone will have zone_id set to NULL (ON DELETE SET NULL)
        zoneRepository.delete(zone);
    }

    // ========== LAYOUT ==========
    
    /**
     * Lädt komplettes Layout für Store
     * 
     * Inkludiert:
     * - Slot-Daten (code, capacity, active)
     * - Position (x, y, width, height)
     * - Belegung (occupiedCount)
     * - Zone
     * 
     * N+1 vermieden durch Eager Loading + Batch-Query
     */
    @Transactional(readOnly = true)
    public List<DhlShelfSlotLayoutDto> getLayout(Long storeId) {
        log.debug("Loading layout for store {}", storeId);
        
        // 1. Layouts mit Slots + Zones laden (1 Query mit Joins)
        List<DhlShelfSlotLayout> layouts = layoutRepository.findByStoreIdWithSlotAndZone(storeId);
        
        if (layouts.isEmpty()) {
            return List.of();
        }
        
        // 2. OccupiedCounts für alle Slots in einem Batch-Query laden
        List<Long> slotIds = layouts.stream()
            .map(l -> l.getShelfSlot().getId())
            .collect(Collectors.toList());
        
        List<Object[]> occupiedCountsRaw = parcelRepository.countByStoreIdAndStatusGroupedBySlot(
            storeId, 
            DhlParcelStatus.STORED, 
            slotIds
        );
        
        Map<Long, Long> occupiedCounts = occupiedCountsRaw.stream()
            .collect(Collectors.toMap(
                row -> (Long) row[0],  // slotId
                row -> (Long) row[1]   // count
            ));
        
        // 3. DTOs bauen
        return layouts.stream()
            .map(layout -> {
                Long slotId = layout.getShelfSlot().getId();
                int occupied = occupiedCounts.getOrDefault(slotId, 0L).intValue();
                return DhlShelfSlotLayoutDto.fromEntity(layout, occupied);
            })
            .collect(Collectors.toList());
    }
    
    /**
     * Batch-Update für Drag&Drop
     * 
     * WICHTIG:
     * - Nur Positionen/Größen werden geändert
     * - shelfSlot.id bleibt stabil
     * - Paketzuordnungen bleiben unverändert
     * - Capacity bleibt unverändert
     */
    @Transactional
    public void updateLayoutBatch(Long storeId, DhlLayoutUpdateRequest request) {
        log.info("Batch-updating {} layouts for store {}", request.getUpdates().size(), storeId);
        
        for (LayoutPositionUpdate update : request.getUpdates()) {
            // Find existing layout
            DhlShelfSlotLayout layout = layoutRepository.findByStoreIdAndSlotId(storeId, update.getSlotId())
                .orElseThrow(() -> new IllegalArgumentException(
                    "Layout for slot " + update.getSlotId() + " not found or access denied"
                ));
            
            // Update position/size
            layout.setGridX(update.getGridX());
            layout.setGridY(update.getGridY());
            layout.setGridWidth(update.getGridWidth());
            layout.setGridHeight(update.getGridHeight());
            
            // Update zone (optional)
            if (update.getZoneId() != null) {
                DhlZone zone = zoneRepository.findByStoreIdAndId(storeId, update.getZoneId())
                    .orElseThrow(() -> new IllegalArgumentException("Zone not found"));
                layout.setZone(zone);
            } else {
                layout.setZone(null);
            }
            
            layoutRepository.save(layout);
        }
        
        log.info("Batch-update completed for store {}", storeId);
    }
    
    /**
     * Erstellt neuen Slot MIT Layout
     * 
     * Atomic: Slot + Layout in einer Transaktion
     */
    @Transactional
    public DhlShelfSlotLayoutDto createSlotWithLayout(Long storeId, DhlCreateSlotWithLayoutRequest request) {
        log.info("Creating slot '{}' with layout for store {}", request.getCode(), storeId);
        
        // Validation
        if (request.getCode() == null || request.getCode().trim().isEmpty()) {
            throw new IllegalArgumentException("Slot code is required");
        }
        if (request.getCapacity() == null || request.getCapacity() < 1) {
            throw new IllegalArgumentException("Capacity must be >= 1");
        }
        
        Store store = storeRepository.findById(storeId)
            .orElseThrow(() -> new IllegalArgumentException("Store not found"));
        
        // Check duplicate code
        if (slotRepository.existsByStoreIdAndCode(storeId, request.getCode())) {
            throw new IllegalArgumentException("Slot with code '" + request.getCode() + "' already exists");
        }
        
        // 1. Create Slot
        DhlShelfSlot slot = new DhlShelfSlot();
        slot.setStore(store);
        slot.setCode(request.getCode());
        slot.setCapacity(request.getCapacity());
        slot.setDescription(request.getDescription());
        slot.setActive(true);
        
        // sortOrder = max + 1
        Integer maxSortOrder = slotRepository.findMaxSortOrderByStoreId(storeId);
        slot.setSortOrder(maxSortOrder != null ? maxSortOrder + 1 : 1);
        
        slot = slotRepository.save(slot);
        
        // 2. Create Layout
        DhlShelfSlotLayout layout = new DhlShelfSlotLayout();
        layout.setStore(store);
        layout.setShelfSlot(slot);
        layout.setGridX(request.getGridX() != null ? request.getGridX() : 0);
        layout.setGridY(request.getGridY() != null ? request.getGridY() : 0);
        layout.setGridWidth(request.getGridWidth() != null ? request.getGridWidth() : 1);
        layout.setGridHeight(request.getGridHeight() != null ? request.getGridHeight() : 1);
        
        if (request.getZoneId() != null) {
            DhlZone zone = zoneRepository.findByStoreIdAndId(storeId, request.getZoneId())
                .orElseThrow(() -> new IllegalArgumentException("Zone not found"));
            layout.setZone(zone);
        }
        
        layout = layoutRepository.save(layout);
        
        return DhlShelfSlotLayoutDto.fromEntity(layout, 0);
    }
}

File: storeFrontend/src/app/core/services/dhl-layout.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  DhlZone,
  DhlZoneRequest,
  DhlShelfSlotLayout,
  DhlLayoutUpdateRequest,
  DhlCreateSlotWithLayoutRequest
} from '@app/core/models/dhl.model';

/**
 * DHL Phase 3A – API Service
 * 
 * HTTP Service für alle 7 Backend-Endpoints.
 * Multi-Tenant: storeId ist immer Teil der URL.
 * 
 * Backend API Base: /api/stores/{storeId}/dhl
 */
@Injectable({
  providedIn: 'root'
})
export class DhlLayoutService {
  private readonly baseUrl = `${environment.apiUrl}/api/stores`;

  constructor(private http: HttpClient) {}

  // ========== ZONES ==========

  /**
   * GET /api/stores/{storeId}/dhl/zones
   * 
   * Lädt alle Zonen eines Stores
   */
  getZones(storeId: number): Observable<DhlZone[]> {
    return this.http.get<DhlZone[]>(`${this.baseUrl}/${storeId}/dhl/zones`);
  }

  /**
   * POST /api/stores/{storeId}/dhl/zones
   * 
   * Erstellt neue Zone
   */
  createZone(storeId: number, request: DhlZoneRequest): Observable<DhlZone> {
    return this.http.post<DhlZone>(`${this.baseUrl}/${storeId}/dhl/zones`, request);
  }

  /**
   * PUT /api/stores/{storeId}/dhl/zones/{zoneId}
   * 
   * Aktualisiert Zone
   */
  updateZone(storeId: number, zoneId: number, request: DhlZoneRequest): Observable<DhlZone> {
    return this.http.put<DhlZone>(`${this.baseUrl}/${storeId}/dhl/zones/${zoneId}`, request);
  }

  /**
   * DELETE /api/stores/{storeId}/dhl/zones/{zoneId}
   * 
   * Löscht Zone (nur wenn keine Slots zugeordnet)
   */
  deleteZone(storeId: number, zoneId: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${storeId}/dhl/zones/${zoneId}`);
  }

  // ========== LAYOUT ==========

  /**
   * GET /api/stores/{storeId}/dhl/layout
   * 
   * Lädt komplettes Layout für Store.
   * 
   * Backend optimiert: N+1-Prevention via Batch-Query.
   * Response enthält: Slots, Positions, Zones, Occupied Counts.
   */
  getLayout(storeId: number): Observable<DhlShelfSlotLayout[]> {
    return this.http.get<DhlShelfSlotLayout[]>(`${this.baseUrl}/${storeId}/dhl/layout`);
  }

  /**
   * PUT /api/stores/{storeId}/dhl/layout
   * 
   * Batch-Update für Drag&Drop.
   * 
   * WICHTIG:
   * - Verändert nur gridX, gridY, gridWidth, gridHeight, zoneId
   * - Niemals: slotId, capacity, parcel-assignments
   * - Transaktion: Alle Updates oder keiner
   * 
   * Verwendung:
   * Nach Drag&Drop im Expertenmodus → Änderungen sammeln → 1x senden
   */
  updateLayoutBatch(storeId: number, request: DhlLayoutUpdateRequest): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${storeId}/dhl/layout`, request);
  }

  /**
   * POST /api/stores/{storeId}/dhl/layout/slots
   * 
   * Erstellt neuen Slot MIT Layout (atomic).
   * 
   * Backend erstellt in einer Transaktion:
   * 1. DhlShelfSlot (code, capacity, description)
   * 2. DhlShelfSlotLayout (gridX, gridY, gridWidth, gridHeight, zoneId)
   * 
   * Response: Komplettes DhlShelfSlotLayout inkl. slotId
   */
  createSlotWithLayout(storeId: number, request: DhlCreateSlotWithLayoutRequest): Observable<DhlShelfSlotLayout> {
    return this.http.post<DhlShelfSlotLayout>(`${this.baseUrl}/${storeId}/dhl/layout/slots`, request);
  }

  // ========== HELPER METHODS ==========

  /**
   * Findet Slot im Layout-Array
   */
  findSlotById(layouts: DhlShelfSlotLayout[], slotId: number): DhlShelfSlotLayout | undefined {
    return layouts.find(layout => layout.slotId === slotId);
  }

  /**
   * Findet Slot im Layout-Array by Code
   */
  findSlotByCode(layouts: DhlShelfSlotLayout[], code: string): DhlShelfSlotLayout | undefined {
    return layouts.find(layout => layout.slotCode.toUpperCase() === code.toUpperCase());
  }

  /**
   * Berechnet Grid-Bounds (für Auto-Scroll)
   */
  getGridBounds(layouts: DhlShelfSlotLayout[]): { maxX: number; maxY: number } {
    if (layouts.length === 0) {
      return { maxX: 0, maxY: 0 };
    }

    const maxX = Math.max(...layouts.map(l => l.gridX + l.gridWidth));
    const maxY = Math.max(...layouts.map(l => l.gridY + l.gridHeight));

    return { maxX, maxY };
  }

  /**
   * Gruppiert Layouts nach Zone
   */
  groupByZone(layouts: DhlShelfSlotLayout[]): Map<string, DhlShelfSlotLayout[]> {
    const groups = new Map<string, DhlShelfSlotLayout[]>();

    layouts.forEach(layout => {
      const key = layout.zoneName || 'Unassigned';
      if (!groups.has(key)) {
        groups.set(key, []);
      }
      groups.get(key)!.push(layout);
    });

    return groups;
  }
}
